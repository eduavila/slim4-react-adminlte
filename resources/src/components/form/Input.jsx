import React from 'react';
import { toCssClasse } from '../../helpers/utils';

const Input = ({
  field, // { name, value, onChange, onBlur }
  form: { touched, errors }, // also values, setXXXX, handleXXXX, dirty, isValid, status, etc.
  ...props
}) => {

  return (
    <div className={`form-group ${toCssClasse(props.cols || `12`)} has-feedback ${touched[field.name] && errors[field.name] ? "has-error" : ""}`}>
      <label className="control-label">
        {props.description}
        {props.isRequired &&
          <span style={{ color: 'red' }} title="Campo ObrigatÃ³rio"><b>*</b></span>}
      </label>

      <input {...field} {...props} className={`form-control ${touched[field.name] && errors[field.name] ? "is-invalid" : ""}`} />

      { touched[field.name] &&
        errors[field.name] && <div className="help-block">{errors[field.name]}</div> }
    </div>
  )
};

export default Input;