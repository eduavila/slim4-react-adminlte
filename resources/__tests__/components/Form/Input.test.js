import React from 'react';
import { shallow, mount } from 'enzyme';
///import { render, fireEvent, wait } from "@testing-library/react";
import toJson from 'enzyme-to-json';
import Input from '../../../src/components/form/Input';
import { updateWrapper, waait } from '../../setup/helpersTest';

const updateField = (wrapper, name, value) => {
    wrapper.simulate('change', {
      persist: () => {},
      target: {
        name,
        value,
      },
    })
  }
const diveFind = (wrapper,query) => wrapper.find(query);


  
describe('<Input />', () => {

    const mockSetFieldValue = jest.fn();
    const mockSetFieldTouched = jest.fn();
    const mockSetFieldOnChange = jest.fn();

    const mockProps = {
        field:{
            name: 'input_mock',
            value: "Teste",
            onChange: mockSetFieldOnChange
        },
        form: { 
            touched: {}, 
            errors: {},
            setFieldValue: mockSetFieldValue, 
            setFieldTouched: mockSetFieldTouched
        },
        props:{
            descrition: 'Campo de mock:'
        }
    }

    describe('render()', () => {
        test('renders the component', () => {
            const wrapper = shallow(<Input { ...mockProps }/>);
        
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        test('Altera valor input', async () => {
            const wrapper = mount(<Input { ...mockProps }/>);

            //Procura input e aplica evento change.
            diveFind(wrapper,'input[name="input_mock"]')
                .simulate('change', {
                    persist: () => {},
                    target: { name: 'input_mock', value: 'teste' }, // invalid
                });
            
            expect(mockSetFieldOnChange.mock.calls.length).toEqual(1);
            expect(mockSetFieldOnChange.mock.calls[0][0].target.value).toEqual('teste');
        });

        test('Msg de erro', async () => {
            const form = {  
                errors: { input_mock: "Teste error" },
                touched: { input_mock: true }
            };
            const wrapper = mount(<Input {...mockProps} form={form} />);

            //Procura input e aplica evento change.
            const input = diveFind(wrapper,'input[name="input_mock"]');
            input.simulate('focus');
            input.simulate('blur');

            const text = wrapper.find('.help-block').text();
            expect(text).toEqual(form.errors.input_mock);
        });
    });
});
