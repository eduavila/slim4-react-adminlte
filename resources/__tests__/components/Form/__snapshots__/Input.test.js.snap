// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> render() renders the component 1`] = `
<div
  className="form-group  col-xs-12 has-feedback "
>
  <label
    className="control-label"
  />
  <input
    className="form-control "
    name="input_mock"
    onChange={[MockFunction]}
    props={
      Object {
        "descrition": "Campo de mock:",
      }
    }
    value="Teste"
  />
</div>
`;
